lib/run_qsimh.h

```diff
@@ -39,46 +39,55 @@ struct QSimHRunner final {
    * @param param Options for parallelism and logging. Also specifies the size
    *   of the 'prefix' and 'root' sections of the lattice.
    * @param maxtime Maximum number of time steps to run.
+   * @param circuit The circuit to be simulated.
    * @param parts Lattice sections to be simulated.
-   * @param gates List of all gates in the circuit.
    * @param bitstrings List of output states to simulate, as bitstrings.
    * @param results Output vector of amplitudes. After a successful run, this
    *   will be populated with amplitudes for each state in 'bitstrings'.
    * @return True if the simulation completed successfully; false otherwise.
    */
+  template <typename Circuit>
   static bool Run(const Parameter& param, unsigned maxtime,
-                  const std::vector<unsigned>& parts,
-                  const std::vector<Gate>& gates,
+                  const Circuit& circuit, const std::vector<unsigned>& parts,
                   const std::vector<uint64_t>& bitstrings,
                   std::vector<std::complex<fp_type>>& results) {
+    if (circuit.num_qubits != parts.size()) {
+      IO::errorf("parts size is not equal to the number of qubits.");
+      return false;
+    }
+
     double t0 = 0.0;
 
     if (param.verbosity > 0) {
       t0 = GetTime();
     }
 
-    HybridData hd = HybridSimulator::SplitLattice(parts, gates);
+    HybridData hd;
+    bool rc = HybridSimulator::SplitLattice(parts, circuit.gates, hd);
 
-    if (hd.num_gatexs < param.num_prefix_gatexs + param.num_root_gatexs) {
-      IO::errorf("error: num_prefix_gates (%u) plus num_root gates (%u) is "
-                 "greater than num_gates_on_the_cut (%u).\n",
-                 param.num_prefix_gatexs, param.num_root_gatexs, hd.num_gatexs);
-      return false;
-    }
+    if (rc) {
```

Comment:
Style suggestion: put the short part of the conditional in the brackets to minimize indentation:
```
if (!rc) {
  return false;
}
// ...rest of function...
```

Reply:
Done.